### PR TITLE
Refactor notmuch window queries, improve documentation, and fix #2900

### LINKED ISSUE #2900: using nm_query_window_duration setting results in time window duration +1
## Expected Behaviour

yesterday i decided to give the nm_query_window_* settings a shot. i like the idea but the result was somewhat unexpected.
i used the following settings:

```
set nm_query_type = "threads"
set nm_query_window_duration = 1
set nm_query_window_timebase = "month"
bind index < vfolder-window-backward
bind index > vfolder-window-forward
```

i expect to see 1 months worth of emails when viewing a notmuch tag.
when using '<' and '>' to navigate from one month to the next, i expect to see 1 month of emails.

## Actual Behaviour

i always see 2 months worth of emails.
when viewing in april i see the following emails:
beginning of march to end of april

navigate back one time window:
beginning of february to end of march

navigate back one time window:
beginning of january to end of february

hence i see every month in duplicate when using the vfolder-window-* keys to navigate into the past.

to get the results i expected using notmuch search, i need to use:
```notmuch search date:0months..``` to list all emails in 'this month'.
this is also true for the other time units such as week. unfortunately it is impossible to set ```nm_query_window_duration = 0``` as that disables the feature.

when setting ```nm_query_window_duration = 2``` i see 3 months worth of emails and navigating 'back in time' moves 2 months at a time.

perhaps there is a setting i should have set or perhaps i misunderstood something in the manual.

## Steps to Reproduce

```
set nm_query_type = "threads"
set nm_query_window_duration = 1
set nm_query_window_timebase = "month"
bind index < vfolder-window-backward
bind index > vfolder-window-forward
```

## How often does this happen?

- Always

## When did it start to happen?

- When I changed my config

```
set nm_query_type = "threads"
set nm_query_window_duration = 1
set nm_query_window_timebase = "month"
bind index < vfolder-window-backward
bind index > vfolder-window-forward
```

## NeoMutt Version

```
NeoMutt 20210205
Copyright (C) 1996-2020 Michael R. Elkins and others.
NeoMutt comes with ABSOLUTELY NO WARRANTY; for details type 'neomutt -vv'.
NeoMutt is free software, and you are welcome to redistribute it
under certain conditions; type 'neomutt -vv' for details.

System: Linux 5.11.10-gentoo-dist (x86_64)
ncurses: ncurses 6.2.20200212 (compiled with 6.2.20200212)
GPGME: 1.14.0
OpenSSL: OpenSSL 1.1.1j  16 Feb 2021
libnotmuch: 5.2.0
storage: gdbm, bdb, lmdb

Configure options: --prefix=/usr --build=x86_64-pc-linux-gnu --host=x86_64-pc-linux-gnu --mandir=/usr/share/man --infodir=/usr/share/info --datadir=/usr/share --sysconfdir=/etc --localstatedir=/var/lib --docdir=/usr/share/doc/neomutt-20210205 --with-sysroot=/ --libdir=/usr/lib64 CCACHE=none --full-doc --enable-nls --enable-notmuch --enable-autocrypt --enable-gpgme --disable-pgp --disable-smime --enable-bdb --enable-gdbm --disable-kyotocabinet --disable-qdbm --disable-tokyocabinet --disable-idn --disable-gss --enable-lmdb --enable-sasl --with-ui=ncurses --sysconfdir=/etc/neomutt --enable-ssl --disable-gnutls --disable-testing

Compilation CFLAGS: -march=skylake -O2 -pipe -std=c99 -fno-delete-null-pointer-checks -D_ALL_SOURCE=1 -D_GNU_SOURCE=1 -D__EXTENSIONS__ -I//usr/include -DNCURSES_WIDECHAR -I/include -I//usr/include/

Default options:
  +attach_headers_color +compose_to_sender +compress +cond_date +debug 
  +encrypt_to_self +forgotten_attachments +forwref +ifdef +imap +index_color 
  +initials +limit_current_thread +multiple_fcc +nested_if +new_mail +nntp +pop 
  +progress +quasi_delete +regcomp +reply_with_xorig +sensible_browser +sidebar 
  +skip_quoted +smtp +status_color +timeout +tls_sni +trash 

Compile options:
  +autocrypt +bkgdset +color +curs_set +fcntl -flock -fmemopen +futimens 
  +getaddrinfo -gnutls +gpgme -gss +hcache -homespool -idn +inotify 
  -locales_hack -lua +meta -mixmaster +nls +notmuch +openssl -pgp +regex +sasl 
  -smime +sqlite +start_color +sun_attachment +typeahead 
MAILPATH="/var/mail"
PKGDATADIR="/usr/share/neomutt"
SENDMAIL="/usr/sbin/sendmail"
SYSCONFDIR="/etc/neomutt"
```

## Extra Info

* Operating System and its version: gentoo up to date as of 2021-04-17
* Were you using multiple copies of NeoMutt at once?: no
* Were you using 'screen' or 'tmux'?: no
* Is your email local (maildir) or remote (IMAP)?: local

### PR DESCRIPTION
**Pull request by Austin-Ray**
This PR is a bit larger than I'd like, but I got ahead of myself. It's best understood by reading each commit message. However, there are four goals in this PR:

 - Refactor window query logic into two components, high-level and low-level, then add tests to protect against regressions.
 - Improve documentation of the feature
 - Fix #2900
 - Don't break any existing configurations so we're not burning our users' good will

Please note that commit 5b01866 is shown out of order on GitHub, it is the **last** commit in this series. I reordered the branch so the refactor comes before fixing #2900 for easier reverting if need be.

---

@neomutt/notmuch-users Please test this and let me know if I've broken a use-case, I've tried my best to test manually, but errors are always possible. If you can contribute a unit test for any broken use-case, I'd greatly appreciate it!

### CODE REVIEW

File: notmuch/config.c [hunk 1]
```diff
@@ -61,6 +62,32 @@ static int nm_default_url_validator(const struct ConfigSet *cs, const struct Con
...
+static int nm_query_window_timebase_validator(const struct ConfigSet *cs,
+                                              const struct ConfigDef *cdef,
+                                              intptr_t value, struct Buffer *err)
+{
+  const char *timebase = (const char *) value;
+  if (!nm_query_window_check_timebase(timebase))
+  {
+    mutt_buffer_printf(
+        err, _("Invalid nm_query_window_timebase value (valid values are: "
+               "hour, day, week, month or year)"));
```
Comment (flatcap) [inline notmuch/config.c:84]: This needs some care.
I suggest quoting the valid strings, then adding an `L10N` comment for the translators,
as it's crucial that they don't translate 'hour', 'day', etc.

An alternative to a `DT_STRING`, is a `DT_ENUM`, currently only used by `$mbox_type`
https://github.com/neomutt/neomutt/blob/cd15d4e6ec3467d36c9cd6cc146ea6a64bb94267/mutt_config.c#L411
https://github.com/neomutt/neomutt/blob/cd15d4e6ec3467d36c9cd6cc146ea6a64bb94267/mx.c#L88-L101

**Review by flatcap — COMMENTED** [review]
I'll have a close look tomorrow

File: notmuch/config.c (hunk 1 above)
Comment (Austin-Ray) [reply notmuch/config.c:84]: > This needs some care.
I suggest quoting the valid strings, then adding an L10N comment for the translators, as it's crucial that they don't translate 'hour', 'day', etc.

Do you have an example code segment that I could look at for correctness? I actually lifted that string from `notmuch.c`.

>  An alternative to a DT_STRING, is a DT_ENUM, currently only used by $mbox_type

That's actually what I mentioned on IRC for a "good first issue".  It's an easy enough change for someone who isn't super familiar with the codebase to do.

File: notmuch/query.c [hunk 2]
```diff
@@ -119,3 +119,112 @@ enum NmQueryType nm_string_to_query_type_mapper(const char *str)
 
   return NM_QUERY_TYPE_UNKNOWN;
 }
+
+/**
+ * query_window_check_timebase - Checks if a given timebase string is valid
```
[the remark this answers is not in the document]
Comment (Austin-Ray) [reply notmuch/query.c]: fixup'd

File: notmuch/config.c (hunk 1 above)
Comment (flatcap) [reply notmuch/config.c:84]: > don't translate 'hour'

Thinking about it a bit more, perhaps we should go for a more enum-compatible error message.
How about:
```
nm_query_window_timebase: value must be one of: hour, day, week, month, year
```

We can code this error as:

```c
  /* L10N: Warning message when setting an enumeration, e.g.
     "mbox_type: value must be one of: mbox, mmdf, mh, maildir" */
  _("%s: value must be one of: %s"), "nm_query_window_timebase", "hour, day, week, month, year"
```

The `L10N` comment is a free-form instruction that's copied verbatim to the `po/*.po` files.
The config name and comma-separated values are not translated.

> "good first issue"

Ah!  Good :-)
And a good second issue would be to improve the error message for enums
by automatically generating the warning message above.

Comment (Austin-Ray) [reply notmuch/config.c:84]: > Thinking about it a bit more, perhaps we should go for a more enum-compatible error message.
> The config name and comma-separated values are not translated.

While the values aren't translated, how do we convey the equivalent time concept to a non-English speaker? Or should we assume some level of English understanding for non-English speakers?

For `mbox_type`, its values are all technical terms so the enum message works perfectly. However, `nm_query_window_timebase` should let the user know that week means, well, a week even though we only accept the English word for it.

Our existing translation aren't error proof in their implementation. Translations, such as German (below) and Japanese, include accepted English values with translated values next to them while others (such as Czech and Spanish) translate the values.

German:
```
Ungültiger Wert für nm_query_window_timebase (gültige Werte: hour (Stunde), day (Tag), week (Woche), month (Monat), year (Jahr))
```

Comment (Austin-Ray) [reply notmuch/config.c:84]: After further thought on the enumerations translations, we can include a `L10N` comment for concepts requiring translation but keeping accepted values untranslated.

For example:
```
/* L10N: Warning message when setting an enumeration.
   Please kept untranslated values alongside translation e.g. hour (Stunde), day (Tag), week (Woche), month (Monat), year (Jahr)*/
```

That seems a perfect use-case for `L10N` comment calling focus to an edge-case in translation.

**Comment by flatcap** [comment]
I rebased this PR and made a few minor non-code fixes.
Unfortunately, I can't update your branch.

I suspect you unticked the box "**Allow edits by maintainers**"
It's at the bottom of the right-hand sidebar.

**Comment by Austin-Ray** [comment]
> a few minor non-code fixes.

What changes are needed? I can make them

**Comment by flatcap** [comment]
I've pushed the changes to [[devel/issue-2900]](https://github.com/neomutt/neomutt/tree/devel/issue-2900)

**Comment by Austin-Ray** [comment]
Updated to modifications in that branch

**Comment by Austin-Ray** [comment]
Anything required for this PR?

**Comment by Austin-Ray** [comment]
Ping

**Review by flatcap — APPROVED** [review]
Sorry this has taken so long.

**Comment by flatcap** [comment]
I can't update this branch, so I've rebased and merged it locally.
Next time, **please** create the branch on the neomutt repo,
or at least ensure that the "Enable maintainer updates" tickbox is ticked.

**Pull request closed without merging**